### ISSUE TITLE
CUTLASS 3.5.1 makes Flash Attention 3 slower?

### ISSUE THREAD

**Issue by fno2010** [Dao-AILab/flash-attention#1289]
Hi all, I just built the latest Flash Attention 3 ([c1d146c](https://github.com/Dao-AILab/flash-attention/commit/c1d146cbd5becd9e33634b1310c2d27a49c7e862)) with the latest CUTLASS release (`v3.5.1`, i.e., [f7b19de](https://github.com/NVIDIA/cutlass/commit/f7b19de32c5d1f3cedfc735c2849f12b537522ee)). But to compare with the build with the revision ([756c351](https://github.com/NVIDIA/cutlass/tree/756c351b4994854b2f8c6dded3821ebbb580876b)) of the CUTLASS submodule suggested by the flash-attention repo, the forward throughput of the `v3.5.1` build is unexpectedly slower in the "causal=False, hdim=128" case.

| ![fav3_cutlass_diff_iscausal_hdim64](https://github.com/user-attachments/assets/10562f5e-3afb-4d6c-a62c-80eee061e68d) | ![fav3_cutlass_diff_iscausal_hdim128](https://github.com/user-attachments/assets/639ba8e7-7a27-4b8a-a2c2-fc9b3971d404) |
|:--|:--|
| ![fav3_cutlass_diff_nocausal_hdim64](https://github.com/user-attachments/assets/0ddf80fd-242e-4c72-b3c9-a7c48908f3d3) | ![fav3_cutlass_diff_nocausal_hdim128](https://github.com/user-attachments/assets/d88421ce-6785-4ee8-8870-88fa80b51411) |

The benchmark is run on H100 for BF16.

I see the CUTLASS v3.5.1 has merged the changes from [756c351](https://github.com/NVIDIA/cutlass/tree/756c351b4994854b2f8c6dded3821ebbb580876b). So it is unexpected. I'm not sure if I missed any compilation option? I just followed the README and used the default options by `python setup.py install`. Does anyone have any idea about this?

**Comment by tridao** [comment]
Yes 3.5.1 changes `DefaultCopy` to not vectorize. We'd need to explicitly vectorize it. This work is under way. You can check out the `tdd` branch (https://github.com/Dao-AILab/flash-attention/tree/tdd) to benchmark. We'll merge it soon.
So far we observe best perf with CUDA 12.3.

**Comment by fno2010** [comment]
> Yes 3.5.1 changes `DefaultCopy` to not vectorize. We'd need to explicitly vectorize it. This work is under way. You can check out the `tdd` branch (https://github.com/Dao-AILab/flash-attention/tree/tdd) to benchmark. We'll merge it soon. So far we observe best perf with CUDA 12.3.

@tridao Thanks for the quick response. I just tested the `tdd` branch. It performs perfectly. Looking forward to seeing it merged. 👍 

And yes, we are indeed using CUDA 12.3 right now.

**Comment by fno2010** [comment]
I believe the `tdd` branch refactors the whole flash 3 codebase and makes the structure clean. But if I understand it correctly, to address the slowness issue that I mentioned above, the only needed change is to add the `-DCUTE_SM90_EXTENDED_MMA_SHAPES_ENABLED` compilation option, right?

**Comment by tridao** [comment]
Yes I think so
